### ISSUE TITLE
Question about the training code

### ISSUE THREAD

**Issue by LZL-CS** [ActiveVisionLab/DFNet#7]
Hi, 
1. I am confused about why here choose to freeze BN layers:

https://github.com/ActiveVisionLab/DFNet/blob/1389760f770851a77e601af1312f19fe065bd185/script/train.py#L111

2. Why use different models:
![image](https://user-images.githubusercontent.com/54640259/211569497-c7781965-dc0a-408f-b733-ffe1f6ca53d2.png)

thanks in advance!

**Comment by chenusc11** [comment]
Hi, this is quite important. 

Since feat_model.eval() is previously set, the batch norm is frozen automatically by PyTorch. 

model's BN layer needs to be frozen manually, since it still needs to update the weights of other layers. This ensures more stable training of the model.

**Comment by LZL-CS** [comment]
Hi @chenusc11, thanks for your reply. But I am still a little bit confused about this code:
<img width="667" alt="image" src="https://user-images.githubusercontent.com/54640259/211735629-0001aca6-f81b-4a6b-95d6-a9ed5751c4ce.png">

the **model** here is just used to return the **pose_** of the input image using DFNet, but the regression module of DFNet for camera pose doesn't use BN Layer at all, so why need to freeze its BN Layer manually?

**Comment by LZL-CS** [comment]
Hi @chenusc11, thanks for your reply. But I am still a little bit confused about this code:

<img width="671" alt="image" src="https://user-images.githubusercontent.com/54640259/211740484-f3d1ed4e-87e5-4a46-ae28-3925969cea99.png">

the **model** here is just used to return the **pose_** of the input image using DFNet, but the regression module of DFNet for camera pose doesn't use BN Layer at all, so why need to freeze its BN Layer manually?

> DFNet(
  (encoder): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): ReLU(inplace=True)
    (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (3): ReLU(inplace=True)
    (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (5): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (6): ReLU(inplace=True)
    (7): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): ReLU(inplace=True)
    (9): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (10): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): ReLU(inplace=True)
    (12): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (13): ReLU(inplace=True)
    (14): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (15): ReLU(inplace=True)
    (16): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (17): Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (18): ReLU(inplace=True)
    (19): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (20): ReLU(inplace=True)
    (21): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (22): ReLU(inplace=True)
    (23): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (24): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (25): ReLU(inplace=True)
    (26): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (27): ReLU(inplace=True)
    (28): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (29): ReLU(inplace=True)
    (30): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (adaptation_layers): AdaptLayers(
    (adapt_layer_0): Sequential(
      (0): Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1))
      (1): ReLU()
      (2): Conv2d(64, 128, kernel_size=(5, 5), stride=(1, 1), padding=(2, 2))
      (3): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (adapt_layer_1): Sequential(
      (0): Conv2d(256, 64, kernel_size=(1, 1), stride=(1, 1))
      (1): ReLU()
      (2): Conv2d(64, 128, kernel_size=(5, 5), stride=(1, 1), padding=(2, 2))
      (3): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (adapt_layer_2): Sequential(
      (0): Conv2d(512, 64, kernel_size=(1, 1), stride=(1, 1))
      (1): ReLU()
      (2): Conv2d(64, 128, kernel_size=(5, 5), stride=(1, 1), padding=(2, 2))
      (3): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (avgpool): AdaptiveAvgPool2d(output_size=1)
  (fc_pose): Linear(in_features=512, out_features=12, bias=True)
)

**Comment by chenusc11** [comment]
Oh yeah. Good catch. Thanks. I think you are right.
It is probably there because I tried some other backbone instead of VGG in some of my experiments, which had BN layer in it. You can remove it if you are using VGG backbone.

**Comment by LZL-CS** [comment]
> Oh yeah. Good catch. Thanks. I think you are right. It is probably there because I tried some other backbone instead of VGG in some of my experiments, which had BN layer in it. You can remove it if you are using VGG backbone.

Hi, got it and thanks for your reply. 

1. I think the code blew has a similar problem (line 432-435): https://github.com/ActiveVisionLab/DFNet/blob/1389760f770851a77e601af1312f19fe065bd185/script/feature/direct_feature_matching.py#L432
Where the optimizer for loss is used for DFNet(dm) rather than pre-trained NeRF-W model, so it's unnecessary to set whole NeRF-W model to requires_grad=False.
https://github.com/ActiveVisionLab/DFNet/blob/1389760f770851a77e601af1312f19fe065bd185/script/train.py#L127

2. The DFnet(dm) in trained based on pre-trained DFNet. Specifically, when train the DFnet(dm), we need to frozen the weights of feature_map BN Layer (pre-trained on DFNet), while continue to update the weights of pose regression module. I think the DFnet(dm) is more like a fine-tune model based on pre-trained DFNet, and so its performace is better than DFnet.

**Comment by chenusc11** [comment]
Hi, yes. I think you are correct! Thanks for the discovery! Will fix this.

On the 2nd point. Yes, I think you are also correct. One of the intuitions behind this could be shown in one of my early experiments below.

<img width="1043" alt="Screen Shot 2023-02-20 at 5 00 10 PM" src="https://user-images.githubusercontent.com/8814381/220165339-253344ca-18a0-4cb7-87e2-00e5c8336981.png">

As you can see. APR can easily overfit the training set pretty well, but it doesn't generalize on the testing sequences. Therefore, no matter what you do to further improve the training on the training set, it will improve very little testing sequence result. Thus I think both RVS and dm experiments would be a good idea to improve APR because
1. RVS generates more training views, therefore covers more area of the scene. like LENS [22]
2. DM enables to further fine-tune DFNet to get better results.

**Comment by LZL-CS** [comment]
> Hi, yes. I think you are correct! Thanks for the discovery! Will fix this.
> 
> On the 2nd point. Yes, I think you are also correct. One of the intuitions behind this could be shown in one of my early experiments below.
> 
> <img alt="Screen Shot 2023-02-20 at 5 00 10 PM" width="1043" src="https://user-images.githubusercontent.com/8814381/220165339-253344ca-18a0-4cb7-87e2-00e5c8336981.png">
> 
> As you can see. APR can easily overfit the training set pretty well, but it doesn't generalize on the testing sequences. Therefore, no matter what you do to further improve the training on the training set, it will improve very little testing sequence result. Thus I think both RVS and dm experiments would be a good idea to improve APR because
> 
> 1. RVS generates more training views, therefore covers more area of the scene. like LENS [22]
> 2. DM enables to further fine-tune DFNet to get better results.

Hi @chenusc11, thanks for your reply.
